### ISSUE TITLE
About  AUTOMATIC METRICS

### ISSUE THREAD

**Issue by starlesschan** [stonyhu/DanceRevolution#11]
Could you release the code of automatic metrics?

**Comment by stonyhu** [comment]
@starlesschan The automatic metrics are the model-based metrics except for the beat coverage and beat hit rate. For FID, ACC, Diversity and Multimodality, we follow [Dancing to Music](https://arxiv.org/pdf/1911.02001.pdf) to train a style classifier on the pose data, which is a MLP and outputs 3 labels, i.e., ballet, hiphop and pop. Then, we use this pre-trained classifier to calculate these four automatic metrics. For beat coverage and beat hit rate, we follow [Yalta et al., 2019](https://github.com/Fhrozen/motion_dance) to calculate. We are archiving the code of style classifier and will share in the repo later.

**Comment by starlesschan** [comment]
> @starlesschan The automatic metrics are the model-based metrics except for the beat coverage and beat hit rate. For FID, ACC, Diversity and Multimodality, we follow [Dancing to Music](https://arxiv.org/pdf/1911.02001.pdf) to train a style classifier on the pose data, which is a MLP and outputs 3 labels, i.e., ballet, hiphop and pop. Then, we use this pre-trained classifier to calculate these four automatic metrics. For beat coverage and beat hit rate, we follow [Yalta et al., 2019](https://github.com/Fhrozen/motion_dance) to calculate. We are archiving the code of style classifier and will share in the repo later.

thanks.

**Comment by starlesschan** [comment]
> @starlesschan The automatic metrics are the model-based metrics except for the beat coverage and beat hit rate. For FID, ACC, Diversity and Multimodality, we follow [Dancing to Music](https://arxiv.org/pdf/1911.02001.pdf) to train a style classifier on the pose data, which is a MLP and outputs 3 labels, i.e., ballet, hiphop and pop. Then, we use this pre-trained classifier to calculate these four automatic metrics. For beat coverage and beat hit rate, we follow [Yalta et al., 2019](https://github.com/Fhrozen/motion_dance) to calculate. We are archiving the code of style classifier and will share in the repo later.

thanks your reply.
Another question, there may be some errors in the part of preprocess code. In the process of extracting acoutistic featrues, missing an important parameter 'hop_length', the ture 'hop_length' parameter is 1024 but the default value is 512, actually resulting the frame of music and dance is not aligned

**Comment by stonyhu** [comment]
@starlesschan The dance pose sequence is 30 fps, such as data/json/ballet_1min/0000_00. When extracting acoustic features, the hop_length is default 512 and thus extracted audio frame sequence is also 30 fps. After that, the paired pose sequence and audio frame sequence is divided into two subsequences with 15 fps. Hence, this is equal to hop_length=1024 (15400/1024=15 fps). Please read the preprocess code carefully.

**Comment by starlesschan** [comment]
> > @starlesschan The automatic metrics are the model-based metrics except for the beat coverage and beat hit rate. For FID, ACC, Diversity and Multimodality, we follow [Dancing to Music](https://arxiv.org/pdf/1911.02001.pdf) to train a style classifier on the pose data, which is a MLP and outputs 3 labels, i.e., ballet, hiphop and pop. Then, we use this pre-trained classifier to calculate these four automatic metrics. For beat coverage and beat hit rate, we follow [Yalta et al., 2019](https://github.com/Fhrozen/motion_dance) to calculate. We are archiving the code of style classifier and will share in the repo later.
> 
> thanks your reply.
> Another question, are there may be some errors in the part of preprocess code? In the process of extracting acoutistic featrues, missing an important parameter 'hop_length', the 'hop_length' parameter should be 1024 but the default value is 512, whether resulting the frame of music and dance not aligned?

Oh, I see. I found it  there has this part in the preprocess code, but why do you want to do like this?
`        new_musics.append([musics[i][j] for j in range(min_seq_len) if j%2==0])
        new_musics.append([musics[i][j] for j in range(min_seq_len) if j%2==1])
        
        new_dances.append([dances[i][j] for j in range(min_seq_len) if j%2==0])
        new_dances.append([dances[i][j] for j in range(min_seq_len) if j%2==1])`

**Comment by starlesschan** [comment]
> @starlesschan The dance pose sequence is 30 fps, such as data/json/ballet_1min/0000_00. When extracting acoustic features, the hop_length is default 512 and thus extracted audio frame sequence is also 30 fps. After that, the paired pose sequence and audio frame sequence is divided into two subsequences with 15 fps. Hence, this is equal to hop_length=1024 (15400/1024=15 fps). Please read the preprocess code carefully.

thanks your reply,  it's very helpful for me

**Comment by starlesschan** [comment]
> @starlesschan The dance pose sequence is 30 fps, such as data/json/ballet_1min/0000_00. When extracting acoustic features, the hop_length is default 512 and thus extracted audio frame sequence is also 30 fps. After that, the paired pose sequence and audio frame sequence is divided into two subsequences with 15 fps. Hence, this is equal to hop_length=1024 (15400/1024=15 fps). Please read the preprocess code carefully.

thanks your work, i helps me a lot , but I still have few problems  when i use the model , could you give some advices ?
1. I use the original dataset to get a trained model,  it performs well at mostly samples, but there are still cases of poor performance in a small sample, such as some samples still  "freezing" randomly, and i don't konw how to slove this problem 
2.I added 2D keypoints of AIST++ dataset to the original dataset for training and I add a gradient reveser layer to eliminate the effects of different dataset distributions, but the model is not stable and performs bad in a lot of samples.
3.Is lstm decoder too sensitive? Have you tried other decoders, such as Transformer Decoder?

**Comment by starlesschan** [comment]
> @starlesschan The dance pose sequence is 30 fps, such as data/json/ballet_1min/0000_00. When extracting acoustic features, the hop_length is default 512 and thus extracted audio frame sequence is also 30 fps. After that, the paired pose sequence and audio frame sequence is divided into two subsequences with 15 fps. Hence, this is equal to hop_length=1024 (15400/1024=15 fps). Please read the preprocess code carefully.

thanks your work, i helps me a lot , but I still have few problems  when i use the model , could you give some advices ?
1. I use the original dataset to get a trained model,  it performs well at mostly samples, but there are still cases of poor performance in a small sample, such as some samples still  "freezing" randomly, and i don't konw how to slove this problem 
2.I added 2D keypoints of AIST++ dataset to the original dataset for training and I add a gradient reveser layer to eliminate the effects of different dataset distributions, but the model is not stable and performs bad in a lot of samples.
3.Is lstm decoder too sensitive? Have you tried other decoders, such as Transformer Decoder?

> @starlesschan The dance pose sequence is 30 fps, such as data/json/ballet_1min/0000_00. When extracting acoustic features, the hop_length is default 512 and thus extracted audio frame sequence is also 30 fps. After that, the paired pose sequence and audio frame sequence is divided into two subsequences with 15 fps. Hence, this is equal to hop_length=1024 (15400/1024=15 fps). Please read the preprocess code carefully.

thanks your work, i helps me a lot , but I still have few problems  when i use the model , could you give some advices ?

1. I use the original dataset to get a trained model,  it performs well at mostly samples, but there are still cases of poor performance in a small sample, such as some samples still  "freezing" randomly, and i don't konw how to slove this problem
2. I added 2D keypoints of AIST++ dataset to the original dataset for training and I add a gradient reveser layer to eliminate the effects of different dataset distributions, but the model is not stable and performs bad in a lot of samples.
3. Is lstm decoder too sensitive? Have you tried other decoders, such as Transformer Decoder?

**Comment by stonyhu** [comment]
> 1. I use the original dataset to get a trained model, it performs well at mostly samples, but there are still cases of poor performance in a small sample, such as some samples still "freezing" randomly, and i don't konw how to slove this problem

**Comment by stonyhu** [comment]
A1: This is the normal phenomenon, since the generation process sometimes would fall into the local optimum and cannot jump out from it, which leads to the "freezing" samples. You can try to generate multiple dance motion sequences for each of those songs. We have released the inference pipeline and you can refer to it.

A2: For the training on AIST++ dataset, I suggest you try different hyperparameter combinations, especially for lambda_v, sliding_window_size and some hidden sizes. This would help you to train a better model on this dataset.

A3: No. This is due to the high spatio-temporal complexity of dance human motions. We have tried the original transformer decoder at the early stage of this work, its results are not desirable. Directly utilizing transformer decoder to predict human motions ignores the temporal-spatial dependencies between pose key joints, which have proven to be highly effective in the state-of-the-art methods in human motion prediction. This issue has been addressed by recent works such as [Ci et al., ICCV19](https://openaccess.thecvf.com/content_ICCV_2019/papers/Ci_Optimizing_Network_Structure_for_3D_Human_Pose_Estimation_ICCV_2019_paper.pdf) and [Aksan et al., arXiv20](https://arxiv.org/pdf/2004.08692.pdf)
